cluster-autoscaler/utils/kubernetes/client.go

```diff
@@ -0,0 +1,56 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package kubernetes
+
+import (
+	"net/url"
+
+	"k8s.io/autoscaler/cluster-autoscaler/config"
+
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/clientcmd"
+	"k8s.io/klog/v2"
+)
+
+// GetKubeConfig returns the rest config from AutoscalingOptions.
+func GetKubeConfig(opts config.AutoscalingOptions) *rest.Config {
```

Comment:
Passing all `AutoscalingOptions` here seems like overkill for this: it only requires a few values. WDYT about moving all kube-client related fields to a dedicated struct and only passing that struct here? If it is necessary in `AutoscalingOptions`, it can still be there.

Reply:
```go
type AutoscalingOptions struct {
    	...

    	// Kubernetes master location.
	Kubernetes string
	// Path to kube configuration if available
	KubeConfigPath string
	// Content type of requests sent to APIServer.
	KubeAPIContentType string
	// Burst setting for kubernetes client
	KubeClientBurst int
	// QPS setting for kubernetes client
	KubeClientQPS float64

    	...
}
```

Yeah I think it's very reasonable. How about moving these fields to a dedicated struct `KubeClientOptions` in *utils/kubernetes/client.go*？

I want to create a kube client in cloudprovider(see  https://github.com/kubernetes/autoscaler/pull/5820#discussion_r1349697340) but `buildCloudProvider` only accept `AutoscalingOptions`:

https://github.com/kubernetes/autoscaler/blob/e7bf3ec8405a2e87cc81ed8779cc8b447ada10cd/cluster-autoscaler/cloudprovider/builder/builder_all.go#L90-L90

If we move kube client releated fields and functions to *utils/kubernetes/client.go*, I'm wondering how to create kube client in cloudprovider.

Reply:
I think the struct should still live next to `AutoscalingOptions` - this package has almost no dependencies and I'd like to keep it that way. Otherwise it would depend on the new file. (And transitively depend on anything else the new file imports.)

What I'm proposing is:

```go
type AutoscalingOptions struct {
	...
	KubeClient KubeClientOptions
	...
}

type KubeClientOptions struct {
 	// Kubernetes master location.
	Kubernetes string
 	// Path to kube configuration if available
 	KubeConfigPath string
 	// Content type of requests sent to APIServer.
 	ContentType string
 	// Burst setting for kubernetes client
 	BurstLimit int
 	// QPS setting for kubernetes client
 	QPSLimit float64
}
```

And then in the new package:

```go
package client

...

func New(opts config.KubeClientOptions) kube_client.Interface {
	...
}
```

With that, given `AutoscalingOptions`, you can still use the new utils to create kube client. Does that make sense?